referenceframe/urdf.go

```diff
@@ -0,0 +1,299 @@
+		} `xml:"geometry"`
+	} `xml:"collision"`
+}
+
+type UrdfJoint struct {
+	XMLName xml.Name `xml:"joint"`
+	Name    string   `xml:"name,attr"`
+	Type    string   `xml:"type,attr"`
+	Origin  struct {
+		XMLName xml.Name `xml:"origin"`
+		RPY     string   `xml:"rpy,attr"` // Fixed frame angle "r p y" format, in radians
+		XYZ     string   `xml:"xyz,attr"` // "x y z" format, in meters
+	} `xml:"origin"`
+	Parent struct {
+		XMLName xml.Name `xml:"parent"`
+		Link    string   `xml:"link,attr"`
+	} `xml:"parent"`
+	Child struct {
+		XMLName xml.Name `xml:"child"`
+		Link    string   `xml:"link,attr"`
+	} `xml:"child"`
+	Axis struct {
+		XMLName xml.Name `xml:"axis"`
+		XYZ     string   `xml:"xyz,attr"` // "x y z" format, in meters
+	} `xml:"axis"`
+	Limit struct {
+		XMLName xml.Name `xml:"limit"`
+		Lower   float64  `xml:"lower,attr"` // translation limits are in meters, revolute limits are in radians
+		Upper   float64  `xml:"upper,attr"` // translation limits are in meters, revolute limits are in radians
+	} `xml:"limit"`
+}
+
+// ParseURDFFile will read a given file and parse the contained URDF XML data into an equivalent ModelConfig struct.
+func ParseURDFFile(filename, modelName string) (Model, error) {
+	//nolint:gosec
+	xmlData, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to read URDF file")
+	}
+	return ConvertURDFToConfig(xmlData, modelName)
+}
+
+// ConvertURDFToConfig will transfer the given URDF XML data into an equivalent ModelConfig. Direct unmarshaling in the
+// same fashion as ModelJSON is not possible, as URDF data will need to be evaluated to accommodate differences
+// between the two kinematics encoding schemes.
+func ConvertURDFToConfig(xmlData []byte, modelName string) (Model, error) {
+	// empty data probably means that the read URDF has no actionable information
+	if len(xmlData) == 0 {
+		return nil, ErrNoModelInformation
+	}
+
+	mc := &ModelConfig{}
+	urdf := &URDFConfig{}
+	err := xml.Unmarshal(xmlData, urdf)
+	if err != nil {
+		return nil, errors.Wrap(err, "Failed to convert URDF data to equivalent URDFConfig struct")
+	}
+
+	if modelName == "" {
+		modelName = urdf.Name
+	}
+
+	parentMap := map[string]string{}
+
+	// Migrate URDF elements into an equivalent ModelConfig representation
+	mc.Name = modelName
+	mc.KinParamType = "SVA"
+
+	// Handle joints
+	for _, jointElem := range urdf.Joints {
+		// Checking for reserved names in this or adjacent elements
+		if jointElem.Name == World {
+			return nil, errors.New("Joints with the name 'world' are not supported by config parsers")
+		}
+
+		// Relationship tracking
+		parentMap[jointElem.Name] = jointElem.Parent.Link
+		parentMap[jointElem.Child.Link] = jointElem.Name
+
+		// Set up the child link mentioned in this joint; fill out the details in the link parsing section later
+		childLink := JsonLink{ID: jointElem.Child.Link, Parent: jointElem.Name}
+
+		switch jointElem.Type {
+		case "continuous", "revolute", "prismatic":
```

Comment:
Can you make sure all joint types are covered by unit tests? It looks like "prismatic" is not.

Reply:
I added an additional URDF file that represents an X-Y gantry composed of only prismatic joints and added parsing of that to the current URDF tests.